Caltech-IPAC/firefly
# DM-10830: Optimize the performance of multi-trace chart updates

https://jira.lsstcorp.org/browse/DM-10830
Explores ways to improve scatter chart performance, especially in Firefox.
To test:
- build firefly in this branch
- goto http://localhost:8080/firefly/demo/ffapi-highlevel-charttest.html
  - edit line 155 of this file from 'radius: 300' -> 'radius: 1000' to get 4k rows

Issues to consider:
- plot update instead of redraw did not increase performance enough, especially on larger plots.
- scattergl support is not complete.
  - latest update v1.28.2 released just a few days ago
  - known bugs:  https://github.com/plotly/plotly.js/issues/130
     - affects us:
        - reversed axis not supported
        - yaxis right+log does not work all the time.
        - zoom select on color map with size very sluggish.
        - 'open' symbol types not working.  ie.  'circle-open'
- what should we do?

Below is a summary of measured time for a single row highlight:
```
     500 rows        GL                SVG
     browser    update    redraw    update    redraw
     chrome     <15       150       40-45     50
     firefox    17-22     150       170-200   210

     4k rows        GL                SVG
     browser    update    redraw    update    redraw
     chrome     15-20     180       215       230
     firefox    20-25     180       1200      1400
```

The code is committed with no scattergl and partial update enabled.
To test the various modes yourself, set sessionStorage variables in debugger console:
        _chartRedraw: set to always redraw full plot_
        _scatterGL: set to convert scatter to scattergl_
```
   sessionStorage.setItem('chartRedraw', 'true' | 'false')
   sessionStorage.setItem('scatterGL', 'true' | 'false')
   sessionStorage.clear()

```
